pkg/conversion/conversion_factory.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion
+
+import (
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/conversion"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+)
+
+// CRConverterFactory instantiates converters that are capable of converting custom resources between different API
+// versions. It supports CEL-based conversion rules from APIConversion resources and the "none" conversion strategy.
+type CRConverterFactory struct {
+	nopConverter     conversion.CRConverter
+	getAPIConversion func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIConversion, error)
+}
+
+var _ conversion.Factory = &CRConverterFactory{}
+
+// NewCRConverterFactory returns a CRConverterFactory that supports APIConversion-based conversions and the "none"
+// conversion strategy.
+func NewCRConverterFactory(apiConversionInformer apisinformers.APIConversionClusterInformer) *CRConverterFactory {
+	return &CRConverterFactory{
+		nopConverter: conversion.NewNOPConverter(),
+		getAPIConversion: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIConversion, error) {
+			return apiConversionInformer.Lister().Cluster(clusterName).Get(name)
+		},
+	}
+}
+
+// NewConverter returns the appropriate conversion.Converter based on the CRD. If the CRD identifies as for a wildcard
+// partial metadata request, the nop converter is used. Otherwise, it returns a CEL-based converter if there is an
+// associated APIConversion, a nop converter if the strategy is "none", or an error otherwise.
+func (f *CRConverterFactory) NewConverter(crd *apiextensionsv1.CustomResourceDefinition) (conversion.CRConverter, error) {
```

Comment:
Are we guaranteed that this is only called once per CRD? If not, do we need to share/dedupe the outcome of the deferred converter?

Reply:
Yes, it's called by `getOrCreateServingInfoFor` only if it's not in the storageMap